### PR TITLE
now the -p flag in validate is working

### PR DESCRIPTION
**Pull request by bakatzir**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/21246

## Description
the `-p` flag in validate was not working.
the purpose of the flag, is being able to validate a file outside of a git dir, without a filename which matches our schema.

## Screenshots
```
(py37) ➜  Downloads demisto-sdk validate -p inc1dentfield.json
You are using demisto-sdk 0.4.3.dev2, however version 0.4.2 is available.
You should consider upgrading via "pip install --upgrade demisto-sdk" command.
Starting validating files structure
Not using git, validating file: inc1dentfield.json
Validating inc1dentfield.json
The files are valid
```

```
(py37) ➜  Downloads demisto-sdk validate -p Test_XDR_Playbook.yml 
You are using demisto-sdk 0.4.3.dev2, however version 0.4.2 is available.
You should consider upgrading via "pip install --upgrade demisto-sdk" command.
Starting validating files structure
Not using git, validating file: Test_XDR_Playbook.yml
Validating Test_XDR_Playbook.yml
ERROR:pykwalify.core:validation.invalid
ERROR:pykwalify.core: --- All found errors ---
ERROR:pykwalify.core:["Cannot find required key 'description'. Path: ''", "Cannot find required key 'description'. Path: '/tasks/0/task'", "Cannot find required key 'description'. Path: '/tasks/4/task'", "Cannot find required key 'description'. Path: '/tasks/7/task'", "Cannot find required key 'description'. Path: '/tasks/8/task'"]
Failed: Test_XDR_Playbook.yml failed.
Missing description in root
Playbook conditional task with id:4 has unhandled condition: else
Playbook conditional task with id:7 has unhandled condition: else
The files were found as invalid, the exact error message can be located above
```

```
(py37) ➜  Downloads demisto-sdk validate -p Zscaler_ioc.yml
You are using demisto-sdk 0.4.3.dev2, however version 0.4.2 is available.
You should consider upgrading via "pip install --upgrade demisto-sdk" command.
Starting validating files structure
Not using git, validating file: Zscaler_ioc.yml
Validating Zscaler_ioc.yml
You have added an image in the yml file, please update the package 
Zscaler_ioc.yml Received the following error for proxy validation:
The default value of the proxy parameter should be ''
Zscaler_ioc.yml Received the following error for insecure validation:
The display name of the insecure parameter should be 'Trust any certificate (not secure)'
The files were found as invalid, the exact error message can be located above
```

### CODE REVIEW

File: demisto_sdk/commands/common/hook_validations/structure.py [hunk 1]
```diff
@@ -45,10 +45,10 @@ def __init__(self, file_path, is_new_file=False, old_file_path=None, predefined_
         self.scheme_name = predefined_scheme or self.scheme_of_file_by_path()
         self.file_type = self.get_file_type()
         self.current_file = self.load_data_from_file()
-        if is_new_file:
+        if is_new_file or predefined_scheme:
             self.old_file = {}
         else:
-            self.old_file = get_remote_file(old_file_path if old_file_path else file_path)
+            self.old_file = get_remote_file(old_file_path) if old_file_path else get_remote_file(file_path)
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/structure.py]: why did you change that?

File: demisto_sdk/commands/common/hook_validations/structure.py [hunk 2]
```diff
@@ -63,7 +63,6 @@ def is_valid_file(self):
             self.is_valid_scheme(),
             self.is_file_id_without_slashes(),
         ]
-
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/structure.py]: restore new line

File: demisto_sdk/commands/validate/file_validator.py [hunk 3]
```diff
@@ -246,37 +246,43 @@ def is_valid_release_notes(self, file_path):
...
 
         In case we encounter an invalid file we set the self._is_valid param to False.
 
         Args:
             modified_files (set): A set of the modified files in the current branch.
+            file_type (str): Used only with -p flag (the type of the file).
         """
+        # print(modified_files)
+        # print(file_type)
+        # sys.exit(0)
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: remove

File: demisto_sdk/commands/validate/file_validator.py [hunk 4]
```diff
@@ -246,37 +246,43 @@ def is_valid_release_notes(self, file_path):
...
                 self._is_valid = False
 
             if self.validate_id_set:
                 if not self.id_set_validator.is_file_valid_in_set(file_path):
                     self._is_valid = False
 
-            elif checked_type(file_path, YML_INTEGRATION_REGEXES):
+            elif checked_type(file_path, YML_INTEGRATION_REGEXES) or file_type == 'integration':
                 image_validator = ImageValidator(file_path)
+                image_validator.file_path = file_path if file_type else image_validator.file_path
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: unclear why you need this, add a comment

File: demisto_sdk/commands/validate/file_validator.py [hunk 5]
```diff
@@ -334,26 +340,27 @@ def validate_modified_files(self, modified_files):  # noqa: C901
                     self._is_valid = False
 
             # incident fields and indicator fields are using the same scheme.
-            elif checked_type(file_path, JSON_INDICATOR_AND_INCIDENT_FIELDS):
+            elif checked_type(file_path, JSON_INDICATOR_AND_INCIDENT_FIELDS) or \
+                    file_type in ('incidentfield', 'indicatorfield'):
                 incident_field_validator = IncidentFieldValidator(structure_validator)
-                if not incident_field_validator.is_valid_file():
+                if not incident_field_validator.is_valid_file(validate_rn=False if file_type else True):
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: are we sure that both type of fields has the same validations? looks confusing

File: demisto_sdk/commands/validate/file_validator.py [hunk 6]
```diff
@@ -408,7 +415,7 @@ def validate_added_files(self, added_files):  # noqa: C901
                     self._is_valid = False
 
                 integration_validator = IntegrationValidator(structure_validator)
-                if not integration_validator.is_valid_file(validate_rn=False):
+                if not integration_validator.is_valid_file():
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: why did you remove? added files should not have RN

File: demisto_sdk/commands/validate/file_validator.py [hunk 7]
```diff
@@ -418,7 +425,7 @@ def validate_added_files(self, added_files):  # noqa: C901
                 structure_validator.file_path = yml_path
                 script_validator = ScriptValidator(structure_validator)
 
-                if not script_validator.is_valid_file(validate_rn=False):
+                if not script_validator.is_valid_file():
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: same

File: demisto_sdk/commands/validate/file_validator.py [hunk 8]
```diff
@@ -584,10 +591,11 @@ def is_valid_structure(self):
                 self.validate_all_files()
         else:
             if self.file_path:
-                self.branch_name = self.get_current_working_branch()
-                self.validate_committed_files()
+                print('Not using git, validating file: {}'.format(self.file_path))
+                self.is_backward_check = False  # if not using git, no need for BC checks
+                self.validate_modified_files({self.file_path}, find_type(self.file_path))
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: not sure what the changes are but i'd rather check that file as new (otherwise it is probably already handled in git)

File: demisto_sdk/commands/common/hook_validations/structure.py (hunk 1 above)
Comment (bakatzir) [reply demisto_sdk/commands/common/hook_validations/structure.py]: no need

File: demisto_sdk/commands/validate/file_validator.py (hunk 5 above)
Comment (bakatzir) [reply demisto_sdk/commands/validate/file_validator.py]: I know, but yes.

**Comment by bakatzir** [comment]
@yaakovi I've refactored it completely since i'm going to the `added` and not to the `modified` function. please re review.

**Comment by bakatzir** [comment]
Re run the validations locally on a non git path, was working as expected.

File: demisto_sdk/commands/validate/file_validator.py [hunk 9]
```diff
@@ -336,24 +336,24 @@ def validate_modified_files(self, modified_files):  # noqa: C901
             # incident fields and indicator fields are using the same scheme.
             elif checked_type(file_path, JSON_INDICATOR_AND_INCIDENT_FIELDS):
                 incident_field_validator = IncidentFieldValidator(structure_validator)
-                if not incident_field_validator.is_valid_file():
+                if not incident_field_validator.is_valid_file(validate_rn=True):
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: already the default value? guessing this is leftovers from the original changes

File: demisto_sdk/commands/validate/file_validator.py [hunk 10]
```diff
@@ -393,13 +394,15 @@ def validate_added_files(self, added_files):  # noqa: C901
...
-            elif re.match(PLAYBOOK_REGEX, file_path, re.IGNORECASE):
+            elif re.match(PLAYBOOK_REGEX, file_path, re.IGNORECASE) or file_type == 'playbook':
                 playbook_validator = PlaybookValidator(structure_validator)
                 if not playbook_validator.is_valid_playbook():
                     self._is_valid = False
 
-            elif checked_type(file_path, YML_INTEGRATION_REGEXES):
+            elif checked_type(file_path, YML_INTEGRATION_REGEXES) or file_type == 'integration':
                 image_validator = ImageValidator(file_path)
+                # if file_type(non git path) the image is not in a separate path
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: not sure this is always true

File: demisto_sdk/commands/validate/file_validator.py [hunk 11]
```diff
@@ -437,22 +440,23 @@ def validate_added_files(self, added_files):  # noqa: C901
                     self._is_valid = False
 
             # incident fields and indicator fields are using the same scheme.
-            elif checked_type(file_path, JSON_INDICATOR_AND_INCIDENT_FIELDS):
+            elif checked_type(file_path, JSON_INDICATOR_AND_INCIDENT_FIELDS) or \
+                    file_type in ('incidentfield', 'indicatorfield'):
                 incident_field_validator = IncidentFieldValidator(structure_validator)
-                if not incident_field_validator.is_valid_file():
+                if not incident_field_validator.is_valid_file(validate_rn=False):
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: if we are in git mode, we should validate release notes.
`validate_rn=not file_type`
`file_type` is None -> not using general path file -> should check RN
`file_type` is not None -> using general path file -> should not check RN

relevant for pre-commit hooks, in the build we have another step that will fail on this check.

File: demisto_sdk/commands/validate/file_validator.py [hunk 12]
```diff
@@ -584,10 +588,11 @@ def is_valid_structure(self):
                 self.validate_all_files()
         else:
             if self.file_path:
-                self.branch_name = self.get_current_working_branch()
-                self.validate_committed_files()
+                print('Not using git, validating file: {}'.format(self.file_path))
+                self.is_backward_check = False  # if not using git, no need for BC checks
+                self.validate_added_files({self.file_path}, find_type(self.file_path))
```
Comment (yaakovi) [inline demisto_sdk/commands/validate/file_validator.py]: for keyword argument, better not to depends on order
```suggestion
                self.validate_added_files({self.file_path}, file_type=find_type(self.file_path))
```

File: demisto_sdk/commands/validate/file_validator.py (hunk 10 above)
Comment (bakatzir) [reply demisto_sdk/commands/validate/file_validator.py]: the -p is to test a single file. usually one downloaded from demisto.
what do you think should be done here?
